core/src/main/java/org/elasticsearch/action/fieldstats/FieldStatsRequest.java

```diff
@@ -33,17 +40,111 @@
 
+    public IndexConstraint[] indexConstraints() {
+        return indexConstraints;
+    }
+
+    public void indexConstraints(IndexConstraint[] indexConstraints) {
+        if (indexConstraints == null) {
+            throw new IllegalArgumentException("Specified indexContraints can't be null");
+        }
+        this.indexConstraints = indexConstraints;
+    }
+
+    public void source(BytesReference content) throws IOException {
+        List<IndexConstraint> indexConstraints = new ArrayList<>();
+        List<String> fields = new ArrayList<>();
+        try (XContentParser parser = XContentHelper.createParser(content)) {
+            String fieldName = null;
+            Token token = parser.nextToken();
+            assert token == Token.START_OBJECT;
+            for (token = parser.nextToken(); token != Token.END_OBJECT; token = parser.nextToken()) {
+                switch (token) {
+                    case FIELD_NAME:
+                        fieldName = parser.currentName();
+                        break;
+                    case START_OBJECT:
+                        if ("index_constraints".equals(fieldName)) {
+                            parseIndexContraints(indexConstraints, parser);
+                        } else {
+                            throw new IllegalArgumentException("Unknown field [" + fieldName + "]");
+                        }
+                        break;
+                    case START_ARRAY:
+                        if ("fields".equals(fieldName)) {
+                            while ((token = parser.nextToken()) != XContentParser.Token.END_ARRAY) {
+                                if (token.isValue()) {
+                                    fields.add(parser.text());
+                                }
+                            }
+                        } else {
+                            throw new IllegalArgumentException("Unknown field [" + fieldName + "]");
+                        }
+                        break;
+                    default:
+                        throw new IllegalArgumentException("Unknown token [" + token + "]");
+                }
+            }
+        }
+        this.fields = fields.toArray(new String[fields.size()]);
+        this.indexConstraints = indexConstraints.toArray(new IndexConstraint[indexConstraints.size()]);
+    }
+
+    private void parseIndexContraints(List<IndexConstraint> indexConstraints, XContentParser parser) throws IOException {
+        Token token = parser.currentToken();
+        assert token == Token.START_OBJECT;
+        String field = null;
+        String currentName = null;
+        for (token = parser.nextToken(); token != Token.END_OBJECT; token = parser.nextToken()) {
+            if (token == Token.FIELD_NAME) {
+                field = currentName = parser.currentName();
+            } else if (token == Token.START_OBJECT) {
+                for (token = parser.nextToken(); token != Token.END_OBJECT; token = parser.nextToken()) {
```

Comment:
I don't like that this is reusing the outter token variable...could we have a unique variable for the inner loop?